lib/proto/proto_pipe.c

```diff
@@ -104,20 +109,43 @@ callback_pipe_read(void * cookie, ssize_t slen)
 		goto fail;
 	len = (size_t)slen;
 
-	/* Did a packet read end prematurely? */
-	if ((P->decr) && (len < PCRYPT_ESZ))
-		goto fail;
-
-	/* Encrypt or decrypt the data. */
-	if (P->decr) {
-		if ((P->wlen = proto_crypt_dec(P->inbuf, P->outbuf, P->k)) == -1)
-			goto fail;
-	} else {
-		proto_crypt_enc(P->inbuf, len, P->outbuf, P->k);
-		P->wlen = PCRYPT_ESZ;
+	/* Get data. */
+	inbuf = &P->inbuf[inpos];
+
+	/* Process as many packets as possible. */
+	while ((len > 0) && (outpos <= OUTBUFSIZE - P->full_buflen)) {
```

Comment:
I would write this as
```
/* Process as much data as possible. */
while (len > 0) {
    /* Stop processing if we don't have space for more output. */
    if (outpos + P->full_buflen > OUTBUFLEN)
        break;
   ...
```

It ends up with the same machine code but it seems clearer to me.